pkg/servicescm/servicescm.go

```diff
@@ -125,6 +134,152 @@ func Generate(name, namespace string) (*core.ConfigMap, error) {
 	return servicesConfigMap, nil
 }
 
+// Parse converts ConfigMap data into the objects representing a Windows services ConfigMap schema
+// Returns error if the given data is invalid in structure or contents.
+func Parse(dataFromCM map[string]string) (*data, error) {
+	if len(dataFromCM) != 2 {
+		return nil, errors.New("services ConfigMap should have exactly 2 keys")
+	}
+
+	value, ok := dataFromCM[servicesKey]
+	if !ok {
+		return nil, errors.Errorf("expected key %s does not exist", servicesKey)
+	}
+	services := &[]Service{}
+	if err := json.Unmarshal([]byte(value), services); err != nil {
+		return nil, err
+	}
+
+	value, ok = dataFromCM[filesKey]
+	if !ok {
+		return nil, errors.Errorf("expected key %s does not exist", filesKey)
+	}
+	files := &[]FileInfo{}
+	if err := json.Unmarshal([]byte(value), files); err != nil {
+		return nil, err
+	}
+
+	return newData(services, files)
+}
+
+// validate ensures the given object represents a valid services ConfigMap, made up of only required services and files.
+// The validation also ensures bootstrap services are defined to always start before controller services.
+func (cmData *data) validate() error {
+	if err := validateRequiredContent(cmData.Services, cmData.Files); err != nil {
+		return err
+	}
+	if err := validateDependencies(cmData.Services); err != nil {
+		return err
+	}
+	return validatePriorities(cmData.Services)
+}
+
+// validateRequiredContent ensures that the given slices are comprised of all the required services/files and only these
+func validateRequiredContent(services []Service, files []FileInfo) error {
```

Comment:
Would it make sense to write this once with a more generic param and call it twice (once for services, once for files) to avoid repeated code?

Reply:
Good suggestion, I think when generics are usable this could be done well though.